alternator/ttl.cc

```diff
@@ -113,4 +142,437 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
+// Alternator Streams if enabled (FIXME: explain how we mark the
+// deletion different from user deletes. We don't do it yet.).
+expiration_service::expiration_service(database& db, service::storage_proxy& proxy)
+        : _db(db)
+        , _proxy(proxy)
+{
+    //FIXME: add metrics for the service
+    //setup_metrics();
+}
+
+// The following is_expired() functions all check if an item with the given
+// expiration time has expired, according to the DynamoDB API rules.
+// The rules are:
+// 1. If the expiration time attribute's value is not a number type,
+//    the item is not expired.
+// 2. The expiration time is measured in seconds since the UNIX epoch.
+// 3. If the expiration time is more than 5 years in the past, it is assumed
+//    to be malformed and ignored - and the item does not expire.
+static bool is_expired(gc_clock::time_point expiration_time, gc_clock::time_point now) {
+    return expiration_time <= now &&
+           expiration_time > now - std::chrono::years(5);
+}
+static bool is_expired(const big_decimal& expiration_time, gc_clock::time_point now) {
+    // Convert the big_decimal to an integer. If it doesn't fit, the
+    // conversion will result in the highest number that does fit.
+    // FIXME: what is expiration_time.scale()? Can it ever be not 1?
+    // Add a test that tries 12345e1 as a timestamp!
+    uint32_t t = static_cast<uint32_t>(expiration_time.unscaled_value());
+    // We assume - and the assumption turns out to be correct - that the
+    // epoch of gc_clock::time_point and the one used by the DynamoDB protocol
+    // are the same (the UNIX epoch in UTC). The resolution (seconds) is also
+    // the same.
+    return is_expired(gc_clock::time_point(gc_clock::duration(std::chrono::seconds(t))), now);
+}
+static bool is_expired(const rjson::value& expiration_time, gc_clock::time_point now) {
+    std::optional<big_decimal> n = try_unwrap_number(expiration_time);
+    if (n) {
+        return is_expired(*n, now);
+    }
+    return false;
+}
+
+// expire_item() expires an item - i.e., deletes it as appropriate for
+// expiration - with CL=QUORUM and (FIXME!) in a way Alternator Streams
+// understands it is an expiration event - not a user-initiated deletion.
+static future<> expire_item(service::storage_proxy& proxy,
+                            const service::query_state& qs,
+                            const std::vector<bytes_opt>& row,
+                            const std::vector<unsigned>& pk_columns,
+                            const std::vector<unsigned>& ck_columns,
+                            schema_ptr schema,
+                            api::timestamp_type ts) {
+    // Prepare the row key to delete
+    std::vector<bytes> exploded_pk;
+    for (unsigned c : pk_columns) {
+        const auto& row_c = row[c];
+        if (!row_c) {
+            // This shouldn't happen - all key columns must have values.
+            // But if it ever happens, let's just *not* expire the item.
```

Comment:
suggestion: if it shouldn't happen, let's log an error; if it somehow does happen, it will be easier to pinpoint later

Reply:
I'm a bit worried that if it does happen, it can happen a million times in a loop. I don't remember we have an existing mechanism for rate-limited logging, do we?

Reply:
If the message is going to be identical, syslog implementations are also free to collapse repetitions. In any case, when debugging, excessive logs are much better than no logs

Reply:
But yeah, I think that we only rate-limit some specific messages in our codebase and the support for this is open-coded, there's no generic mechanism for that

Reply:
I'm adding a FIXME for this. I don't know if logging is appropriate without good rate limiting - if anything is worse than suddenly TTL stopping to work it's suddenly the performance tanking because of excessive logging. Another option is to add a metric, but the whole metric thing is also left as a FIXME.